// @slack/socket-mode/dist/src/SlackWebSocket.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.SlackWebSocket = exports.WS_READY_STATES = void 0;
const node_diagnostics_channel_1 = require("node:diagnostics_channel");
const undici_1 = require("undici");
const errors_1 = require("./errors");
const logger_1 = __importStar(require("./logger"));
exports.WS_READY_STATES = ['CONNECTING', 'OPEN', 'CLOSING', 'CLOSED'];
function isPingPongMessage(message) {
    if (typeof message !== 'object' || message === null) {
        return false;
    }
    if (!('websocket' in message && message.websocket instanceof undici_1.WebSocket)) {
        return false;
    }
    if (!('payload' in message && Buffer.isBuffer(message.payload))) {
        return false;
    }
    return true;
}
/**
 * Encapsulates the Slack-specific details around establishing a WebSocket connection with the Slack backend.
 * Manages the ping/pong heartbeat of the connection.
 */
class SlackWebSocket {
    // python equiv: Connection
    static loggerName = 'SlackWebSocket';
    options;
    logger;
    websocket;
    /**
     * The last timetamp that this WebSocket received pong from the server
     */
    lastPongReceivedTimestamp;
    /**
     * Sentinel checking if Slack sent us a close frame or not, in order to be able
     * to terminate underlying socket gracefully.
     */
    closeFrameReceived;
    /**
     * Reference to the timeout timer we use to listen to pings from the server
     */
    serverPingTimeout;
    /**
     * Reference to the timeout timer we use to listen to pongs from the server
     */
    clientPingTimeout;
    openHandler = null;
    errorHandler = null;
    messageHandler = null;
    closeHandler = null;
    pingHandler = null;
    pongHandler = null;
    static pingChannel = (0, node_diagnostics_channel_1.channel)('undici:websocket:ping');
    static pongChannel = (0, node_diagnostics_channel_1.channel)('undici:websocket:pong');
    constructor({ url, client, dispatcher, logger, logLevel = logger_1.LogLevel.INFO, pingInterval = 5000, pingPongLoggingEnabled = false, serverPingTimeoutMS = 30000, clientPingTimeoutMS = 5000, }) {
        this.options = {
            url,
            client,
            dispatcher,
            logLevel,
            pingInterval,
            pingPongLoggingEnabled,
            serverPingTimeoutMS,
            clientPingTimeoutMS,
        };
        if (logger) {
            this.logger = logger;
        }
        else {
            this.logger = logger_1.default.getLogger(SlackWebSocket.loggerName, logLevel);
        }
        this.websocket = null;
        this.closeFrameReceived = false;
    }
    /**
     * Establishes a connection with the Slack backend
     */
    connect() {
        this.logger.debug('Initiating new WebSocket connection.');
        this.websocket = new undici_1.WebSocket(this.options.url, { dispatcher: this.options.dispatcher });
        this.openHandler = () => {
            this.logger.debug('WebSocket open event received (connection established)!');
            this.monitorPingToSlack();
        };
        this.websocket.addEventListener('open', this.openHandler);
        this.errorHandler = (event) => {
            if (!(event instanceof undici_1.ErrorEvent)) {
                this.logger.warn(`Expected ErrorEvent but received ${event.constructor.name} (type: ${event.type})`);
                return;
            }
            this.logger.error(`WebSocket error occurred: ${event.message}`);
            this.disconnect();
            this.options.client.emit('error', new errors_1.SMWebsocketError(event.error ?? new Error(event.message)));
        };
        this.websocket.addEventListener('error', this.errorHandler);
        this.messageHandler = (event) => {
            if (!(event instanceof undici_1.MessageEvent)) {
                this.logger.warn(`Expected MessageEvent but received ${event.constructor.name} (type: ${event.type})`);
                return;
            }
            const isBinary = typeof event.data !== 'string';
            this.options.client.emit('ws_message', event.data, isBinary);
        };
        this.websocket.addEventListener('message', this.messageHandler);
        this.closeHandler = (event) => {
            if (!(event instanceof undici_1.CloseEvent)) {
                this.logger.warn(`Expected CloseEvent but received ${event.constructor.name} (type: ${event.type})`);
                return;
            }
            this.logger.debug(`WebSocket close frame received (code: ${event.code}, reason: ${event.reason})`);
            this.closeFrameReceived = true;
            this.disconnect();
        };
        this.websocket.addEventListener('close', this.closeHandler);
        // Subscribe to undici diagnostics_channel for WebSocket ping/pong frame events.
        // These channels fire for ALL undici WebSocket instances, so we filter by matching instance.
        this.pingHandler = (message) => {
            if (!isPingPongMessage(message)) {
                this.logger.warn('Received unexpected ping diagnostics message format');
                return;
            }
            if (message.websocket !== this.websocket)
                return;
            if (this.options.pingPongLoggingEnabled) {
                this.logger.debug(`WebSocket received ping from Slack server (data: ${message.payload?.toString()})`);
            }
            this.monitorPingFromSlack();
        };
        SlackWebSocket.pingChannel.subscribe(this.pingHandler);
        this.pongHandler = (message) => {
            if (!isPingPongMessage(message)) {
                this.logger.warn('Received unexpected pong diagnostics message format');
                return;
            }
            if (message.websocket !== this.websocket)
                return;
            if (this.options.pingPongLoggingEnabled) {
                this.logger.debug(`WebSocket received pong from Slack server (data: ${message.payload?.toString()})`);
            }
            this.lastPongReceivedTimestamp = Date.now();
        };
        SlackWebSocket.pongChannel.subscribe(this.pongHandler);
    }
    /**
     * Disconnects the WebSocket connection with Slack, if connected.
     */
    disconnect() {
        if (this.websocket) {
            // Disconnecting a WebSocket involves a close frame handshake so we check if we've already received a close frame.
            // If so, we can terminate the underlying socket connection and let the client know.
            if (this.closeFrameReceived) {
                this.logger.debug('Terminating WebSocket (close frame received).');
                this.cleanup();
            }
            else if (this.websocket.readyState === undici_1.WebSocket.CLOSING) {
                // A close frame was already sent but the peer hasn't responded. Force-terminate rather than
                // waiting for the ws library's closeTimeout (~30s) while the ping monitor logs repeated warnings.
                this.logger.debug('Terminating WebSocket (close frame sent but no response, force-terminating).');
                this.cleanup();
            }
            else {
                // If we haven't received a close frame yet, then we send one to the peer, expecting to receive a close frame
                // in response.
                this.logger.debug('Sending close frame (status=1000).');
                this.websocket.close(1000); // 1000 = Normal Closure
            }
        }
        else {
            this.logger.debug('WebSocket already disconnected, flushing remainder.');
            this.cleanup();
        }
    }
    /**
     * Clean up any underlying intervals, timeouts and the WebSocket.
     */
    cleanup() {
        if (this.websocket) {
            if (this.openHandler)
                this.websocket.removeEventListener('open', this.openHandler);
            if (this.errorHandler)
                this.websocket.removeEventListener('error', this.errorHandler);
            if (this.messageHandler)
                this.websocket.removeEventListener('message', this.messageHandler);
            if (this.closeHandler)
                this.websocket.removeEventListener('close', this.closeHandler);
        }
        this.openHandler = null;
        this.errorHandler = null;
        this.messageHandler = null;
        this.closeHandler = null;
        if (this.pingHandler)
            SlackWebSocket.pingChannel.unsubscribe(this.pingHandler);
        if (this.pongHandler)
            SlackWebSocket.pongChannel.unsubscribe(this.pongHandler);
        this.pingHandler = null;
        this.pongHandler = null;
        this.websocket = null;
        clearTimeout(this.serverPingTimeout);
        clearInterval(this.clientPingTimeout);
        // Emit event back to client letting it know connection has closed (in case it needs to reconnect if
        // reconnecting is enabled)
        this.options.client.emit('close');
    }
    /**
     * Returns true if the underlying WebSocket connection is active, meaning the underlying
     */
    isActive() {
        // python equiv: SocketModeClient.is_connected
        if (!this.websocket) {
            this.logger.debug('isActive(): websocket not instantiated!');
            return false;
        }
        this.logger.debug(`isActive(): websocket ready state is ${exports.WS_READY_STATES[this.websocket.readyState]}`);
        return this.websocket.readyState === undici_1.WebSocket.OPEN;
    }
    /**
     * Retrieve the underlying WebSocket readyState. Returns `undefined` if the WebSocket has not been instantiated,
     * otherwise will return a number between 0 and 3 inclusive representing the ready states.
     */
    get readyState() {
        return this.websocket?.readyState;
    }
    /**
     * Sends data via the underlying WebSocket. Accepts an errorback argument.
     */
    send(data, cb) {
        try {
            this.websocket?.send(data);
            cb(undefined);
        }
        catch (err) {
            cb(err);
        }
    }
    /**
     * Confirms WebSocket connection is still active; fires whenever a ping event is received
     * If we don't receive another ping from the peer before the timeout, we initiate closing the connection.
     */
    monitorPingFromSlack() {
        clearTimeout(this.serverPingTimeout);
        this.serverPingTimeout = setTimeout(() => {
            this.logger.warn(`A ping wasn't received from the server before the timeout of ${this.options.serverPingTimeoutMS}ms!`);
            this.disconnect();
        }, this.options.serverPingTimeoutMS);
    }
    /**
     * Monitors WebSocket connection health; sends a ping to peer, and expects a pong within a certain timeframe.
     * If that expectation is not met, we disconnect the websocket.
     */
    monitorPingToSlack() {
        this.lastPongReceivedTimestamp = undefined;
        let pingAttemptCount = 0;
        clearInterval(this.clientPingTimeout);
        this.clientPingTimeout = setInterval(() => {
            const now = Date.now();
            try {
                const pingMessage = `Ping from client (${now})`;
                if (!this.websocket) {
                    this.logger.error('WebSocket not available, skipping ping.');
                    return;
                }
                (0, undici_1.ping)(this.websocket, Buffer.from(pingMessage));
                if (this.lastPongReceivedTimestamp === undefined) {
                    pingAttemptCount += 1;
                }
                else {
                    // if lastPongReceivedTimestamp is defined, then the server has responded to pings at some point in the past
                    pingAttemptCount = 0;
                }
                if (this.options.pingPongLoggingEnabled) {
                    this.logger.debug(`Sent ping to Slack: ${pingMessage}`);
                }
            }
            catch (e) {
                this.logger.error(`Failed to send ping to Slack (error: ${e})`);
                this.disconnect();
                return;
            }
            // default invalid state is: sent > 3 pings to the server and it never responded with a pong
            let isInvalid = pingAttemptCount > 3;
            if (this.lastPongReceivedTimestamp !== undefined) {
                // secondary invalid state is: if we did receive a pong from the server,
                // has the elapsed time since the last pong exceeded the client ping timeout
                const millis = now - this.lastPongReceivedTimestamp;
                isInvalid = millis > this.options.clientPingTimeoutMS;
            }
            if (isInvalid) {
                this.logger.warn(`A pong wasn't received from the server before the timeout of ${this.options.clientPingTimeoutMS}ms!`);
                this.disconnect();
            }
        }, this.options.clientPingTimeoutMS / 3);
        this.logger.debug('Started monitoring pings to and pongs from Slack.');
    }
}
exports.SlackWebSocket = SlackWebSocket;
